instrumentation/kafka/kafka-clients/kafka-clients-0.11/javaagent/src/test/java/io/opentelemetry/instrumentation/kafkaclients/KafkaClientDefaultTest.java

```diff
@@ -0,0 +1,329 @@
+import io.opentelemetry.sdk.trace.data.LinkData;
+import io.opentelemetry.sdk.trace.data.SpanData;
+import io.opentelemetry.semconv.trace.attributes.SemanticAttributes;
+import java.nio.charset.StandardCharsets;
+import java.time.Duration;
+import java.util.Collections;
+import java.util.List;
+import java.util.concurrent.ExecutionException;
+import java.util.concurrent.TimeUnit;
+import java.util.concurrent.TimeoutException;
+import java.util.concurrent.atomic.AtomicReference;
+import org.apache.kafka.clients.consumer.ConsumerRecord;
+import org.apache.kafka.clients.consumer.ConsumerRecords;
+import org.apache.kafka.clients.producer.ProducerRecord;
+import org.assertj.core.api.AbstractLongAssert;
+import org.junit.jupiter.api.DisplayName;
+import org.junit.jupiter.api.Test;
+import org.junit.jupiter.api.extension.RegisterExtension;
+import org.junit.jupiter.params.ParameterizedTest;
+import org.junit.jupiter.params.provider.ValueSource;
+
+public class KafkaClientDefaultTest extends KafkaClientPropagationBaseTest {
+
+  @RegisterExtension
+  static final InstrumentationExtension testing = AgentInstrumentationExtension.create();
+
+  @DisplayName("test kafka produce and consume")
+  @ParameterizedTest(name = "{index} => test headers: {0}")
+  @ValueSource(booleans = {true, false})
+  public void testKafkaProducerAndConsumerSpan(boolean testHeaders) throws Exception {
+    String greeting = "Hello Kafka!";
+    testing.runWithSpan(
+        "parent",
+        () -> {
+          ProducerRecord<Integer, String> producerRecord =
+              new ProducerRecord<>(SHARED_TOPIC, greeting);
+          if (testHeaders) {
+            producerRecord
+                .headers()
+                .add("test-message-header", "test".getBytes(StandardCharsets.UTF_8));
+          }
+          producer
+              .send(
+                  producerRecord,
+                  (meta, ex) -> {
+                    if (ex == null) {
+                      testing.runWithSpan("producer callback", () -> {});
+                    } else {
+                      testing.runWithSpan("producer exception: " + ex, () -> {});
+                    }
+                  })
+              .get(5, TimeUnit.SECONDS);
+        });
+
+    awaitUntilConsumerIsReady();
+    ConsumerRecords<?, ?> records = consumer.poll(Duration.ofSeconds(5).toMillis());
+    assertThat(records.count()).isEqualTo(1);
+
+    // iterate over records to generate spans
+    for (ConsumerRecord<?, ?> record : records) {
+      testing.runWithSpan(
+          "processing",
+          () -> {
+            assertThat(record.value()).isEqualTo(greeting);
+            assertThat(record.key()).isNull();
+          });
+    }
+    testing.waitForTraces(2);
+    AtomicReference<SpanData> producerSpan = new AtomicReference<>();
+    testing.waitAndAssertSortedTraces(
+        orderByRootSpanKind(SpanKind.INTERNAL, SpanKind.CONSUMER),
+        trace -> {
+          producerSpan.set(trace.getSpan(1));
```

Comment:
Nit: can you move this after the `trace.hasSpansSatisfyingExactly()` call? In case there's less than 2 spans in the trace, we want the assertion to fail nicely instead of getting an out of bounds exception.